### ISSUE TITLE
Allow assistant accounts to see draw when public page not enabled

### ISSUE THREAD

**Issue by czlee** [czlee/tabbycat#442]
Currently the "View Draw" button on the assistant account opens the public draw page, which is great, except when the draw isn't made available to the public (as in, the page is not enabled).

**Comment by philipbelesky** [comment]
Hmm, in ca18013c5ed05d209f5f958010fe67fb5667ce0a I've changed it so that assistants see the same Draw Display page that admins do, minus the top row of buttons. 

At the moment though the Display Motions and Print options are behind superuser accounts. Is there much of a problem changing those pages to be accessible by assistants do you think? It seems like in many circumstances they are exposed to motions early anyway and so should be presumed to have that level of confidence?

**Comment by czlee** [comment]
Hmm, this is a good question. I mean, the correct answer is that we should implement #131, but I'm guessing it's not (yet) worth the effort...

Unreleased motions I think I'd to keep to superusers only, different tournaments will have different requirements here and it's better to be conservative. But I can see the case the other way: the _really_ conservative tournaments on motion availability are few in number (basically the major regional titles) and shouldn't even be entering motions into the database before GA release. So I wouldn't be flatly opposed to making motions available to assistants, but then, I'm also not sure what its use would be. Of course, released motions are fair game for everyone.

Printing I'm happy to make available to assistants, it's just exactly what they can already see, but in a different format.

**Comment by philipbelesky** [comment]
The problem is also that if assistants can't see motions then they can't print ballots because some might contain said motion. But then again that probably happens pretty rarely. Maybe we should err on the side of caution and allow assistants to see unreleased draws but not the printable items.

**Comment by czlee** [comment]
Ah. Yeah, I don't know, now that you mention it I can imagine this being a frustrating roadblock for some tournaments. I always advise major tournaments not to enter motions into the database until GA release anyway, so I can see a case for just allowing assistants to see motions. I'd just want to have very obvious warning messages in place.

**Comment by philipbelesky** [comment]
Yea, as a compromise we could perhaps have a single setting in config along the lines of "Allow assistant users to print ballots and display motions (before their public release)?" that defaults to on? I'm not sure how extensive #131 should be, but this would probably be of value to that (or could be easily superseeded).

**Comment by czlee** [comment]
Yeah, that would work.

If we wanted to do it using the authorization framework, roughly the idea would be:
* Create a user group "assistants"
* Define permissions for each thing ("can print ballots", "can access motions", etc.)
* Use `PermissionRequiredMixin` to regulate access to those pages
* Either grant or don't grant the assistants group the relevant permission

Regulating access to those pages would be easy—that's built into the framework. Things that I think would be hard:
* Making a suitable user interface that allows users to add or remove these permissions. Obviously they can do so through edit database, but if we run with the authorization framework, it doesn't fit "neatly" with the dynamic preferences framework.
* Making it per-tournament. I think we'd need to define user groups and permissions per tournament if we want to do this, which is certainly doable, but probably requires some commitment to building infrastructure if it's going to be sustainable.

That's quite a bit. If we make it a dynamic preference, we'd still need some infrastructure—we'd need a mixin analogous to `PublicTournamentPageMixin`, that does the following:
- Redirects to the login page if the user is not authenticated (_i.e._, inherits from `LoginRequiredMixin`)
- Grants permission if the user is a superuser
- If the user is logged in but not a superuser, checks the relevant tournament option, and then redirects to the login page if the option's disabled, and displays the page if it's enabled.

There's work there, but I think it's a lot less work. Basically we inherit from [`UserPassesTestMixin`](https://docs.djangoproject.com/en/1.10/topics/auth/default/#django.contrib.auth.mixins.UserPassesTestMixin) and `TournamentMixin`, and implement a test that checks the tournament option. Which, now that I think about it, is pretty easy.

The main benefit of doing more graduated permissions is the ability to set permissions on a per-user basis. I can see some use-cases—in very large teams, maybe you want to grant access to motions only to certain individuals, or have ballot check-in on a specific account. But even at the likes of WUDC I've never really longed for anything like this. So having spent a few minutes thinking this through, compared to just setting tournament-wide what assistants have access to, I'm becoming increasingly hesitant about the value of #131, and more of the view that your suggested compromise should just be our long-term solution…

Thoughts?

**Comment by philipbelesky** [comment]
I don't think its worth building out something using the authorisations framework ala #131 just yet. Having a preference in the existing settings area just for this case would be 1) much less work and 2) fit within the existing framework of how Tabbycat is configured (ie users dont need to figure out this whole new way of managing users). 

I think the benefit of #131 is if we ever moved to a more centralised model where we didn't want assistants and users having access to every tournament within an instance. Until then having sensible defaults and 1-2 settings for edge cases seems like the easier option.

**Comment by philipbelesky** [comment]
@czlee is this finished now? Or are we keeping it around as a placeholder for the more ambitious user roles features?

**Comment by czlee** [comment]
Umm, I never actually implemented everything, I just put the mixin in place and applied it to one example. We still need to apply it to the other relevant pages.

**Comment by philipbelesky** [comment]
So I've applied the mixin to the other pages in ef5439266ca909d3f73574e10142d30564455e34.

However I changed the implementation slightly as it seemed to only make sense as a hierarchy of access than a collections of booleans. For example being able to print scoresheets implies being able to see motions and to see the draw; whereas feedback sheets only implies the latter. As a result there is just a single choice preference that sets out 3 tiers of access, and each page using the mixin can specify which of the possible choices allows access.

This is a little clumsy, but less so than having to figure out and implement a way to resolve how a clash between a CanSeeDraw preference and a CanSeeMotions preference would work. In the future if there is a need for more complex hierarchies of assistant permissions we could go back to something more resembling the original implementation but this seems effective for now. I'll merge into develop so it's available for Easters, but will leave up to you whether we should close this issue and branch.
